src/modules/iotjs_module_process.c

```diff
@@ -76,11 +76,50 @@ static jerry_value_t wait_for_source_callback(
 }
 
 
+typedef struct Sources {
+  jerry_value_t file_array;
+  struct Sources* next;
+} Sources_t;
+
 // Export JS module received from the debugger client
 JS_FUNCTION(DebuggerGetSource) {
-  jerry_value_t res;
-  jerry_debugger_wait_for_client_source(wait_for_source_callback, NULL, &res);
-  return res;
+  jerry_debugger_wait_for_source_status_t receive_status;
+  uint8_t counter = 0;
+  Sources_t* files = (Sources_t*)iotjs_buffer_allocate(sizeof(Sources_t));
+  Sources_t* head = files;
+  do {
+    jerry_value_t res;
+    receive_status =
+        jerry_debugger_wait_for_client_source(wait_for_source_callback, NULL,
+                                              &res);
+
+    if (receive_status == JERRY_DEBUGGER_SOURCE_RECEIVED) {
+      files->file_array = res;
+      files->next = (Sources_t*)iotjs_buffer_allocate(sizeof(Sources_t));
+      files = files->next;
+      files->next = NULL;
+
+      counter++;
+    }
+
+    if (receive_status == JERRY_DEBUGGER_CONTEXT_RESET_RECEIVED) {
+      iotjs_environment_config(iotjs_environment_get())
+          ->debugger->context_reset = true;
+      break;
+    }
+  } while (receive_status != JERRY_DEBUGGER_SOURCE_END);
+
+  jerry_value_t ret_val = jerry_create_array(counter);
+  counter = 0;
+  Sources_t* previous;
+  while (head->next != NULL) {
+    jerry_set_property_by_index(ret_val, counter++, head->file_array);
+    jerry_release_value(head->file_array);
+    previous = head;
+    head = head->next;
+    iotjs_buffer_release((char*)previous);
+  }
```

Comment:
Could we just put the sources into the array when they are received? I mean have two steps: we create a chain list, then we copy the chain list into an array. I think we can do it in one step.

Reply:
Yes, you are actually right, I just realised the size of `jerry arrays` can be adjusted dynamically.